### DISCUSSION TITLE
Are there any provisions for 3D files such as .OBJ .STL in collectionBuilder

### DISCUSSION THREAD

**Discussion by amh17** [CollectionBuilder/collectionbuilder.github.io#88] [General]
List would include other file types such as :

.FBX .PLY .3DS .VRML .SKP .DAE etc. 

Thinking of the ability to add an in-browser 3D viewer as well as metadata and so forth.

**Comment by evanwill** [comment]
@amh17 there isn't a 3D viewer built in by default yet. However, in CollectionBuilder-CSV it wouldn't be *too* hard to add a 3D viewer item layout. I have done it using two approaches:

- use a 3rd party embed viewer--> using a service such as Sketchfab makes it relatively easy similar to embedding a youtube video (and these platforms may be a good place to share and host your 3d models anyway to facilitate discovery and reuse). basically need to copy their example embed code to set up a custom item display_template, and swap out the iframe src for each item in your collection (that will be the object_location for your items). There is also a few different 3d object embed services that could serve your items similar to platforms like Sketchfab in a more customized/private way.
- add a javascript 3d viewer library--> this might take a bit more advanced customization depending on the viewer. You would add the viewer code via CDN or in assets/lib, create the custom item display_template, and use the object_location of your items to load the correct file. For example, on the relatively simple side, you could use [model-viewer](https://modelviewer.dev/) (which only supports glTF/GLB format, which you should be able to convert your items into) to create something quickly. More complicated libraries might have more features and support for other formats. For example, I have seen [p5js](https://p5js.org/reference/#/p5/loadModel) or [three.js](https://threejs.org/) (more generic 3d model rendering libraries), or [3dhop](https://github.com/cnr-isti-vclab/3DHOP) (a sort of museum specific 3d viewer).

**Reply by evanwill** [reply]
For example, using [model-viewer](https://modelviewer.dev/), a basic model embed for an item page (assuming object_location is a relative path or url to a glb / gltf model file) would look like: 

```
<!-- import model-viewer component from CDN -->
<script type="module" src="https://ajax.googleapis.com/ajax/libs/model-viewer/3.1.1/model-viewer.min.js"></script>
<!-- add 3d viewer to page -->
<div class="ratio" style="--bs-aspect-ratio: 115%;">
    <model-viewer alt="{{ page.title }}" src="{{ page.object_location }}"  shadow-intensity="1" camera-controls touch-action="pan-y"></model-viewer>
</div>
```

**Reply by evanwill** [reply]
If you have come across another 3d viewer, let me know and I can tell you how viable it would be to add to collectionbuilder!

**Reply by evanwill** [reply]
An update just in case someone comes across this, there is a basic implementation of the model-viewer as an [item include option in cb-csv](https://github.com/CollectionBuilder/collectionbuilder-csv/blob/main/_includes/item/3d-model-viewer.html) now.
